src/frontend/handlers/invoice.test.ts

```diff
@@ -51,4 +72,138 @@ describe("invoice handler", () => {
+    );
+    expect(response.status).toBe(200);
+    expect(response.text).toContain(MN_RATES_MISSING.bannerText);
+    expect(response.text).not.toContain(MN_UNEXPECTED_CHARGE.bannerText);
+  });
+
+  test("Page displays unexpected charge message", async () => {
+    // The unexpected charge line should take precedence over the line item that's over threshold.
+    givenExtractResults = `{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"2"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"${MN_UNEXPECTED_CHARGE.magicNumber}"}`;
+    mockedFetchS3.mockResolvedValue(givenExtractResults);
+
+    const request = supertest(app);
+    const response = await request.get(
+      `/invoice?contract_id=${contractId}&year=2023&month=03`
+    );
+    expect(response.status).toBe(200);
+    expect(response.text).toContain(MN_UNEXPECTED_CHARGE.bannerText);
+  });
+
+  test("Page displays above threshold message", async () => {
+    // The single line item that's over threshold should cause the warning
+    // even though the other line items are okay or below threshold.
+    givenExtractResults = `{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"0"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"2"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"-2"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"0"}`;
+    mockedFetchS3.mockResolvedValue(givenExtractResults);
+
+    const request = supertest(app);
+    const response = await request.get(
+      `/invoice?contract_id=${contractId}&year=2023&month=03`
+    );
+    expect(response.status).toBe(200);
+    expect(response.text).toContain("Invoice above threshold");
+    expect(response.text).not.toContain("Invoice below threshold");
+    expect(response.text).not.toContain("Invoice within threshold");
+  });
+
+  test("Page displays below threshold message", async () => {
+    // The single line item that's over threshold should cause the warning
+    // even though the other line items are okay.
+    givenExtractResults = `{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"0"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"-2"}
+{"month":"03","year":"2023","contract_id":"${contractId}","price_difference_percentage":"0"}`;
+    mockedFetchS3.mockResolvedValue(givenExtractResults);
+
+    const request = supertest(app);
+    const response = await request.get(
+      `/invoice?contract_id=${contractId}&year=2023&month=03`
+    );
+    expect(response.status).toBe(200);
+    expect(response.text).toContain("Invoice below threshold");
+    expect(response.text).not.toContain("Invoice within threshold");
+  });
+
+  test("Page displays within threshold message", async () => {
```

Comment:
This `givenExtractResults` looks the same as the one set up in the `beforeEach()` and used in the first test "Page displays invoice info". 
Maybe remove one to avoid duplication?